### ISSUE TITLE
Building from source on Windows?

### ISSUE THREAD

**Issue by slonopotamus** [moby/moby#43304]
**Description**

It isn't clear how to build Moby from source

**Steps to reproduce the issue:**
1. Search for build instructions in https://github.com/moby/moby/blob/master/README.md
2. Search for build instructions in https://github.com/moby/moby/blob/master/CONTRIBUTING.md
3. Search for build instructions in https://github.com/moby/moby/blob/master/TESTING.md

**Describe the results you received:**

Build instructions are nowhere.

**Describe the results you expected:**

I expected to find clear and simple build instructions like "do this, do that, MAGIC, now you have ready-to-use dockerd".

**Additional information you deem important (e.g. issue happens only occasionally):**

I accidentally came across https://github.com/moby/moby/blob/master/hack/README.md but that doesn't help either. It says I should not call `hack\make.ps1` directly, but instead just call `make`. Obviously, there isn't any `make` on Windows.

If I call `hack\make.ps1` directly, it fails to find `windmc` and nothing says where I am supposed to get that from.

**Comment by slonopotamus** [comment]
UPD: I later found https://github.com/moby/moby/blob/master/docs/contributing/software-req-win.md

But its instructions also do not work.

```
> docker build -t nativebuildimage -f .\Dockerfile.windows .
Sending build context to Docker daemon  64.07MB
Step 1/11 : FROM microsoft/windowsservercore
pull access denied for microsoft/windowsservercore, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
```

**Comment by slonopotamus** [comment]
Bonus:

I found `win` target in https://github.com/moby/moby/blob/master/Makefile and tried building in on a Linux box.

That fails with:

```
docker run --rm -i --privileged  -e DOCKER_CROSSPLATFORMS -e BUILD_APT_MIRROR -e BUILDFLAGS -e KEEPBUNDLE -e DOCKER_BUILD_ARGS -e DOCKER_BUILD_GOGC -e DOCKER_BUILD_OPTS -e DOCKER_BUILD_PKGS -e DOCKER_BUILDKIT -e DOCKER_BASH_COMPLETION_PATH -e DOCKER_CLI_PATH -e DOCKER_DEBUG -e DOCKER_EXPERIMENTAL -e DOCKER_GITCOMMIT -e DOCKER_GRAPHDRIVER -e DOCKER_LDFLAGS -e DOCKER_PORT -e DOCKER_REMAP_ROOT -e DOCKER_ROOTLESS -e DOCKER_STORAGE_OPTS -e DOCKER_TEST_HOST -e DOCKER_USERLANDPROXY -e DOCKERD_ARGS -e TEST_FORCE_VALIDATE -e TEST_INTEGRATION_DIR -e TEST_SKIP_INTEGRATION -e TEST_SKIP_INTEGRATION_CLI -e TESTDEBUG -e TESTDIRS -e TESTFLAGS -e TESTFLAGS_INTEGRATION -e TESTFLAGS_INTEGRATION_CLI -e TEST_FILTER -e TIMEOUT -e VALIDATE_REPO -e VALIDATE_BRANCH -e VALIDATE_ORIGIN_BRANCH -e VERSION -e PLATFORM -e DEFAULT_PRODUCT_LICENSE -e PRODUCT -v "/home/marat/moby/bundles:/go/src/github.com/docker/docker/bundles" -v "/home/marat/moby/.git:/go/src/github.com/docker/docker/.git" -v docker-dev-cache:/root/.cache -v docker-mod-cache:/go/pkg/mod/    -t "docker-dev" DOCKER_CROSSPLATFORMS=windows/amd64 hack/make.sh cross
hack/dind: 41: exec: DOCKER_CROSSPLATFORMS=windows/amd64: not found
make: *** [Makefile:245: win] Error 127
```

**Comment by TBBle** [comment]
From memory, you're supposed to pull the appropriate servercore image, and tag it locally as `microsoft/windowsservercore`. Then the instructions at the top of Dockerfile.windows should work.

**Comment by grochoge** [comment]
The instructions at the top of Dockerfile.windows build dockerd but download docker.exe 17.06.2-ce from docker.com.

How do I build the docker.exe?

**Comment by grochoge** [comment]
Ah, that's in a separate repo and needs to be built on Linux and copied over (according to the docs on that repo at least): https://github.com/docker/cli

If it can't be built on Windows, it should be removed from the hack/build.ps1 IMO.

**Comment by colathro** [comment]
As of right now, for others hitting this, general commands to pull the image and tag it would be:

docker pull mcr.microsoft.com/windows/servercore:ltsc2022

docker image tag mcr.microsoft.com/windows/servercore:ltsc2022 microsoft/windowsservercore

**Comment by olljanat** [comment]
Nowadays best way looks to be:
1. Install Go
2. Include these extra binaries https://github.com/moby/moby/blob/34b8670b1a0e0890263b24c3d604445570905dbb/Dockerfile.windows#L229-L236 (I copied them to same folders with Go like it is done in Dockerfile.Windows)
3. Configure environment variables `GOBIN=C:\Program Files\Go\bin` and `GO111MODULE=off`
4. Build with PowerShell command:
```powershell
.\hack\make.ps1 -Daemon -Noisy
```

That way you don't need Docker for build (which makes easier to play with test versions) and after first build it is very rebuild when debugging code.

**Comment by cowwoc** [comment]
For what it's worth, here is what I had to do to build under Windows 11 (Microsoft Windows [Version 10.0.26100.3775]):

1. Install Go for Windows
2. In powershell, run: `mkdir "$ENV:GOPATH/src/github.com/docker"`
3. `cd "$ENV:GOPATH/src/github.com/docker"`
4. `git clone https://github.com/moby/moby.git`
5. `mv moby docker`
6. `cd docker`
7. `go install github.com/tc-hib/go-winres@latest`
8. `$ENV:GO111MODULE="off"`
9. `.\hack\make.ps1 -Daemon -Noisy`
10. `dockerd.exe` will be output into the `bundles` directory relative to the repository root.

Notice that I did not have to download any of the binaries mentioned in `Dockerfile.windows`.

PS: The top of make.ps1 contains helpful comments. For example, you can add `-NoOpt` to build the binary in debug mode for debugging purposes (which is precisely what I am doing).

If you get:

> mkdir C:\ProgramData\docker\exec-root: Cannot create a file when that file already exists.

when trying to run dockerd, it means you need to re-run it as an administrator. I hope this helps others!

**Comment by olljanat** [comment]
Two more recommendations which did just came to my mind:
* Use dedicated Windows Server 2025 (or what ever is latest version when you read this) as development servers instead of Windows 11
  * This is because it default to process isolation mode instead of Hyper-V isolation.
* Enable containerd with configuration described in #47955

**Comment by cowwoc** [comment]
> Two more recommendations which did just came to my mind:
> 
> * Use dedicated Windows Server 2025 (or what ever is latest version when you read this) as development servers instead of Windows 11
>   
>   * This is because it default to process isolation mode instead of Hyper-V isolation.
> * Enable containerd with configuration described in [Windows: Run containerd as managed process #47955](https://github.com/moby/moby/pull/47955) and make sure that it is same version than defined in `Dockerfile.windows`

It's not clear to me why we'd want to do this. Most developers run Windows 11, not Server edition, so this makes the build less accessible.

I personally had no problem building under Windows 11. That problem did you have?

With respect to using containerd, can you please link to a specific comment what configuration changes we need to make? The issue you linked to contains a very long discussion thread so it's not obvious which comment you are referring to.

Thank you.

**Comment by olljanat** [comment]
> Most developers run Windows 11, not Server edition, so this makes the build less accessible.

Yes and of the Docker development happens in Linux but when you are  working on with Windows specific features it is better to be as near of production environment as possible. Windows containers are hard enough anyway so it does not make sense add any extra complexity.

If you are not developing Windows containers specific features/fixes it is better to use Linux or WSL. Example VS Code has extensions for both scenarios so you can do actual coding from Windows.

> I personally had no problem building under Windows 11. That problem did you have?

Building is easy, testing and debugging is the hard part. Those who really want to use Windows 11, you get quite near by running Docker Engine with parameter: `--exec-opt isolation=process`

Look:
* https://docs.docker.com/reference/cli/dockerd/#configure-container-isolation-technology-windows
* https://learn.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/hyperv-container

> With respect to using containerd, can you please link to a specific comment what configuration changes we need to make? The issue you linked to contains a very long discussion thread so it's not obvious which comment you are referring to.

It is not issue but pull request which added that feature. Usage example is in first message and it should works also as command line parameter `--default-runtime io.containerd.runhcs.v1`
And the containers binaries must be in same folder with dockerd.exe
